src/snowflake/connector/network_async.py

```diff
@@ -0,0 +1,637 @@
+                logger.info(f"Session cleanup failed: {e}")
+        self._active_sessions.clear()
+        self._idle_sessions.clear()
+
+
+class SnowflakeRestfulAsync(SnowflakeRestful):
+    def __init__(
+        self,
+        host: str = "127.0.0.1",
+        port: int = 8080,
+        protocol: str = "http",
+        inject_client_pause: int = 0,
+        connection: SnowflakeConnection | None = None,
+    ) -> None:
+        # YICHUAN: We need to do this instead of using asyncio.run() because of session pooling; if we simply use
+        # asyncio.run(), the loops held by aiohttp.ClientSession will be closed, so the next time the cached session
+        # is used it will raise "RuntimeError: Event loop is closed"
+        self._loop_runner = EventLoopThreadRunner()
+        self._loop_runner.start()
+
+        self._host = host
+        self._port = port
+        self._protocol = protocol
+        self._inject_client_pause = inject_client_pause
+        self._connection = connection
+        self._lock_token = Lock()
+        self._sessions_map: dict[
+            str | None, SessionPoolAsync
+        ] = collections.defaultdict(lambda: SessionPoolAsync(self))
+
+        # OCSP mode (OCSPMode.FAIL_OPEN by default)
+        ssl_wrap_socket.FEATURE_OCSP_MODE = (
+            self._connection._ocsp_mode()
+            if self._connection
+            else ssl_wrap_socket.DEFAULT_OCSP_MODE
+        )
+        # cache file name (enabled by default)
+        ssl_wrap_socket.FEATURE_OCSP_RESPONSE_CACHE_FILE_NAME = (
+            self._connection._ocsp_response_cache_filename if self._connection else None
+        )
+
+        # This is to address the issue where requests hangs
+        _ = "dummy".encode("idna").decode("utf-8")
+
+    def close(self) -> None:
+        if hasattr(self, "_token"):
+            del self._token
+        if hasattr(self, "_master_token"):
+            del self._master_token
+        if hasattr(self, "_id_token"):
+            del self._id_token
+        if hasattr(self, "_mfa_token"):
+            del self._mfa_token
+
+        # run_until_complete_safe(self._loop, self.close_async())
+        self._loop_runner.run_coro(self.close_async())
+        self._loop_runner.stop()
+
+    async def close_async(self) -> None:
+        for session_pool in self._sessions_map.values():
+            await session_pool.close_async()
+
+    def fetch(self, *args, **kwargs) -> dict[Any, Any]:
+        # return run_until_complete_safe(self._loop, self.fetch_async(*args, **kwargs))
+        return self._loop_runner.run_coro(self.fetch_async(*args, **kwargs))
+
+    async def fetch_async(
+        self,
+        method: str,
+        full_url: str,
+        headers: dict[str, Any],
+        data: dict[str, Any] | None = None,
+        timeout: int | None = None,
+        **kwargs,
+    ) -> dict[Any, Any]:
+        """Carry out API request with session management."""
+
+        class RetryCtx(TimeoutBackoffCtx):
```

Comment:
nit: we shall reuse the code in the sync code, we can move this class out of the `SnowflakeRestful`